File: week_1/percolation/percolation.rb
require 'test/unit'
require_relative 'weighted_quick_union_uf'

class Percolation

  def initialize(n)
    @size = n
    @top = 0
    @bottom = @size ** 2 + 1
    @qf = WeightedQuickUnionUF.new(@size ** 2 + 2)
    @opened = Array.new(@size){Array.new(@size){false}}
  end

  def open(i, j)
    @opened[i -1][j -1] = true
    
    @qf.union(get_qf_index(i, j), @top) if i == 1
    @qf.union(get_qf_index(i, j), @bottom) if i == @size

    @qf.union(get_qf_index(i, j), get_qf_index(i, j - 1)) if (j > 1 && is_open(i, j - 1))
    @qf.union(get_qf_index(i, j), get_qf_index(i, j + 1)) if (j < @size && is_open(i, j + 1))
    @qf.union(get_qf_index(i, j), get_qf_index(i - 1 , j)) if (i > 1  && is_open(i - 1, j))
    @qf.union(get_qf_index(i, j), get_qf_index(i + 1 , j)) if (i < @size && is_open(i + 1, j))
  end

  def is_open(i, j)
    @opened[i -1][j -1]
  end

  def is_full(i, j)
    return @qf.connected(@top, get_qf_index(i, j)) if (0 < i && i <= @size && 0 < j && j <= @size)
    false
  end

  def percolates
    @qf.connected(@top, @bottom)
  end

  def get_qf_index(i, j)
    @size * (i -1) + j
  end

end

class RandomValues

  class << self
    def generate_for n
      range = (1..n).to_a
      range.product(range).shuffle
    end
  end

end

class PercolationTest < Test::Unit::TestCase
  
  def setup
    n = 5
    @perc = Percolation.new(n)
    @rv = RandomValues.generate_for(n)
  end


  def test_percolates

    c = 0
    
    while @perc.percolates == false do
      @perc.open(*@rv[c])
      c += 1
    end

    print "Pecolates after #{c} attempts"

  end

end

File: week_1/percolation/percolation.py
import unittest
from random import shuffle
from weighted_quick_union_uf import WeightedQuickUnionUF

class Percolation:

  def __init__(self, n):
    self.size = n
    self.top = 0
    self.bottom = self.size ** 2 + 1
    self.qf = WeightedQuickUnionUF(self.size ** 2 + 2)
    self.opened = [[False for x in range(self.size)] for x in range(self.size)]

  def open(self, i, j):
    self.opened[i -1][j -1] = True
    
    if (i == 1):
      self.qf.union(self.get_qf_index(i, j), self.top) 
    elif (i == self.size):
      self.qf.union(self.get_qf_index(i, j), self.bottom) 

    if (j > 1 and self.is_open(i, j - 1)):
      self.qf.union(self.get_qf_index(i, j), self.get_qf_index(i, j - 1)) 
    if (j < self.size and self.is_open(i, j + 1)):
      self.qf.union(self.get_qf_index(i, j), self.get_qf_index(i, j + 1)) 
    if (i > 1  and self.is_open(i - 1, j)):
      self.qf.union(self.get_qf_index(i, j), self.get_qf_index(i - 1 , j))
    if (i < self.size and self.is_open(i + 1, j)):
      self.qf.union(self.get_qf_index(i, j), self.get_qf_index(i + 1 , j)) 

  def is_open(self, i, j):
    self.opened[i -1][j -1]

  def is_full(self, i, j):
    if (0 < i and i <= self.size and 0 < j and j <= self.size):
      return self.qf.connected(self.top, self.get_qf_index(i, j)) 
    else:
      return False

  def percolates(self):
    return self.qf.connected(self.top, self.bottom)

  def get_qf_index(self, i, j):
    return self.size * (i -1) + j
  
class RandomValues:

  def __init__(self, n):
    values = list(range(1, n+1))
    self.sample = []
    for y in values:
      self.sample += [[x,y] for x in values]

class PercolationTest(unittest.TestCase):

  def test_percolation(self):
    n = 5
    perc = Percolation(n)
    rv = RandomValues(n).sample
    print(rv)
    c = 0

    while (perc.percolates() == False):
      print(rv[c])
      perc.open(*rv[c])
      print perc.percolates()
      c += 1
      print perc.qf.array

    print("Pecolates after " + str(c) + " attempts")


if __name__ == '__main__':
  unittest.main()
